File: clients/apps/lib/services/message_service.dart
import 'dart:io';

import 'package:dio/dio.dart';

import '../models/models.dart';
import '../models/requests.dart';
import '../network/message_api_client.dart';

class MessageService {
  final MessageApiClient _client;
  final Dio _dio;

  MessageService(this._client, this._dio);

  Future<List<Message>> getMessages(
    String conversationId, {
    int limit = 50,
  }) async {
    final res = await _client.getMessages(conversationId, limit: limit);
    return res.data.data?.messages ?? [];
  }

  Future<Message> sendMessage(
    String conversationId,
    String content, {
    String? clientMsgId,
  }) async {
    final res = await _client.sendMessage(
      SendMessageRequest(
        conversationId: conversationId,
        content: content,
        clientMsgId: clientMsgId,
      ),
    );
    return res.data.data!;
  }

  Future<Message> sendImageMessage(
    String conversationId,
    File file, {
    String? clientMsgId,
  }) async {
    final fileName = file.uri.pathSegments.isNotEmpty
        ? file.uri.pathSegments.last
        : 'image.jpg';
    final form = FormData.fromMap({
      'conversationId': conversationId,
      if (clientMsgId != null) 'clientMsgId': clientMsgId,
      'file': await MultipartFile.fromFile(file.path, filename: fileName),
    });
    final res = await _dio.post<Map<String, dynamic>>(
      '/messages/images',
      data: form,
      options: Options(
        contentType: 'multipart/form-data',
        sendTimeout: const Duration(seconds: 60),
        receiveTimeout: const Duration(seconds: 60),
      ),
    );
    return Message.fromJson(res.data!['data'] as Map<String, dynamic>);
  }

  Future<Message> toggleReaction(
    String conversationId,
    String messageId,
    String type,
  ) async {
    final res = await _dio.post<Map<String, dynamic>>(
      '/messages/$conversationId/$messageId/reactions',
      data: {'type': type},
    );
    return Message.fromJson(res.data!['data'] as Map<String, dynamic>);
  }

  Future<Message> sendDirectMessage(
    String recipientId,
    String content, {
    String? clientMsgId,
  }) async {
    final res = await _client.sendDirectMessage(
      SendDirectMessageRequest(
        recipientId: recipientId,
        content: content,
        clientMsgId: clientMsgId,
      ),
    );
    return res.data.data!;
  }

  Future<Message> updateMessage(
    String conversationId,
    String messageId,
    String content,
  ) async {
    final res = await _client.updateMessage(
      conversationId,
      messageId,
      UpdateMessageRequest(content: content),
    );
    return res.data.data!;
  }

  Future<void> deleteMessage(String conversationId, String messageId) async {
    await _client.deleteMessage(conversationId, messageId);
  }
}

File: clients/packages/shared/src/services/message.service.ts
import type {
  ApiResponse,
  CreateMessageDTO,
  Message,
  MessagesListResponse,
} from '../types'
import { http } from './http'

export const messageService = {
  async getMessages(
    conversationId: string,
    limit = 50,
    before?: string,
  ): Promise<ApiResponse<MessagesListResponse>> {
    const res = await http.get<ApiResponse<MessagesListResponse>>(
      `/messages/${conversationId}`,
      { params: { limit, before } },
    )
    return res.data
  },

  async sendMessage(data: CreateMessageDTO): Promise<ApiResponse<Message>> {
    const body = {
      conversationId: data.conversationId,
      content: data.content,
      type: data.messageType || 'text',
      ...(data.metadata ? { metadata: data.metadata } : {}),
      ...(data.clientMsgId ? { clientMsgId: data.clientMsgId } : {}),
      ...(data.replyToId ? { replyToId: data.replyToId } : {}),
    }
    const res = await http.post<ApiResponse<Message>>('/messages', body)
    return res.data
  },

  async updateMessage(
    conversationId: string,
    messageId: string,
    content: string,
  ): Promise<ApiResponse<Message>> {
    const res = await http.patch<ApiResponse<Message>>(
      `/messages/${conversationId}/${messageId}`,
      { content },
    )
    return res.data
  },

  async deleteMessage(
    conversationId: string,
    messageId: string,
  ): Promise<ApiResponse<void>> {
    const res = await http.delete<ApiResponse<void>>(
      `/messages/${conversationId}/${messageId}`,
    )
    return res.data
  },

  async toggleReaction(
    conversationId: string,
    messageId: string,
    type: string,
  ): Promise<ApiResponse<Message>> {
    const res = await http.post<ApiResponse<Message>>(
      `/messages/${conversationId}/${messageId}/reactions`,
      { type },
    )
    return res.data
  },

  async sendImageMessage(
    conversationId: string,
    file: File,
    clientMsgId?: string,
    onProgress?: (pct: number) => void,
  ): Promise<ApiResponse<Message>> {
    const form = new FormData()
    form.append('conversationId', conversationId)
    form.append('file', file)
    if (clientMsgId) form.append('clientMsgId', clientMsgId)
    const res = await http.post<ApiResponse<Message>>(
      '/messages/images',
      form,
      {
        headers: { 'Content-Type': 'multipart/form-data' },
        onUploadProgress: (e) => {
          if (onProgress && e.total) onProgress(Math.round((e.loaded / e.total) * 100))
        },
      },
    )
    return res.data
  },

  async sendDirectMessage(
    recipientId: string,
    content: string,
    type = 'text',
    clientMsgId?: string,
  ): Promise<ApiResponse<Message>> {
    const res = await http.post<ApiResponse<Message>>('/messages/direct', {
      recipientId,
      content,
      type,
      ...(clientMsgId ? { clientMsgId } : {}),
    })
    return res.data
  },
}
